### PR TITLE
Proposal for OCI/Docker image repository instead of S3 Object Store

### PR DESCRIPTION
**Pull request by marccampbell**

### CODE REVIEW

File: design/registry-instead-of-s3.md [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+
+## Detailed Design
+
+New CLI flags will be added to `kots install` and `kots pull` to provide a local registry to be used.
+When these flags are not provided, KOTS will install Docker Distribution statefulset and service using a ClusterIP.
+This will be an insecure registry.
+
+
+Application manifests are currently packaged as a `.tar.gz` file.
+KOTS will continue to do this, but use the [ORAS Go Module](https://github.com/deislabs/oras#oras-go-module) to package these into an image.
```
Comment (divolgin) [inline design/registry-instead-of-s3.md:34]: How about adding each file as a layer?  This opens possibility for server side deduplication.

**Comment by divolgin** [comment]
There is no mention of kURL in this.  Would it have to provide a different package/install script based on the chosen version of kotsadm?

File: design/registry-instead-of-s3.md (hunk 1 above)
Comment (marccampbell) [reply design/registry-instead-of-s3.md:34]: Hmm, that feels like a small optimization for a lot of complexity. A big application could have 1000s of files. With most YAML documenting being small, would there be a big win in deduplication?

An interesting alternative is to consider 1 image, where we just add each tgz to it, sequenced. So we could pull, add the sequence as a layer, and then push.

File: design/registry-instead-of-s3.md [hunk 2]
```diff
@@ -0,0 +1,50 @@
+# Optionally replace object store requirement with local registry
+
+KOTS requires an object store in order to store persistent versions of application version artifacts.
+KOTS also supports optionally providing a local OCI/Docker image repository to store application container images.
+This proposal removes the object store requirement in favor of the image repository.
+
+## Goals
+
+We could remove the requirement to have an object store (currently minio or rook) by using an image repository.
```
Comment (laverya) [inline design/registry-instead-of-s3.md:9]: will the image repository not also require an object store?

Comment (laverya) [reply design/registry-instead-of-s3.md:9]: and it seems I misunderstood
externally hosted registry

File: design/registry-instead-of-s3.md (hunk 1 above)
Comment (divolgin) [reply design/registry-instead-of-s3.md:34]: yeah, ok.  good point.  images do have a limit on the number of layers.

**Pull request merged**